README
pod2text Net::GPSD3.pm > README

If this is still here it means the programmer was too lazy to create the readme file.

You can create it now by using the command shown above from this directory.

At the very least you should be able to use this set of instructions
to install the module...

perl Makefile.PL
make
make test
make install

If you are on a windows box you should use 'nmake' rather than 'make'.


Changes
Revision history for Perl module Net::GPSD3

TODO
  - Support SUBFRAME objects better
  - Add Tests for many GPS devices
  - Add a capability for handlers to turn off watcher mode.
  - Consolidate code between Net::GPSD3->watch and Net::GPSD::POE->session
  - Figure out a better way to use the default handlers
  - Review time vs. datetime
    - Figure out if we can drop DateTime due to the load time.
    - Figure out if we can make DateTime a runtime requirement.

0.18 2011-09-26
  - Fixed POLL method GPSD API change - RT 73489 ryan from presslab.us
  - Dropped EXE_FILES section from Makefile.PL - RT 70414 - fschlich at cis.fu-berlin.de 
  - Fixed spelling issues in pod - RT 70410 - fschlich at cis.fu-berlin.de

0.17 2011-04-10
  - Moved Net::GPSD3::POE to separate package

0.16 2011-04-08
  - Updated perl-Net-GPSD3-Handler.pl to use two handlers
  - Updated POE interface based on feedback from Rocco Caputo <rcaputo@pobox.com>
    - Switched from inline_states session to object_states session
    - Added stop and shutdown capabilities
    - Added POE methods pause and resume

0.15 2011-04-05
  - Added POE::Session capability
    - Added Net::GPSD3::POE->session
    - Updated tests
    - Fixed Net::GPSD3->constructor to support inheritance
    - Added POE script example
  - Update script documentation

0.14 2011-03-26
  - Added Documentation and Examples
  - Added poll->tpv alias for poll->fix method
  - Updated support for GST object
  - Updated perl-Net-GPSD3-poll.pl to show much more data
  - Updated default handler changed tpv->time to tpv->timestamp

0.13 2011-03-25
  - Added poll->fix method
  - Added poll->sky method
  - Renamed internal method poll->fixes to _fixes
  - Renamed internal method poll->skyviews to _skyviews
  - Updated POLL documentation
  - Added GST object support

0.12 2011-03-24
  - gpsd 2.96 protocol 3.4 support
    - Do not expect the protocol to stabilize until gpsd 3.00
  - Added support for new POLL interface
  - Added support for SUBFRAME object
  - Updated objects to support new W3C timestamp format
  - Added RPM spec file to the distribution
  - Updated tests for the new protocol 3.4
  - Updated Net/GPSD3.pm
    - Moved defaults into methods
    - Added object caching in gpsd->cache method
    - Removed the intersperse capability as redundant with caching
  - Renamed example scripts
  - Dropped strftime support (not backwards compatible!)

0.11 2010-01-14
  - gpsd protocol 3.3 support
  - Moved to IO::Socket::INET6 to support both IPv4 and IPv6

0.10 2010-01-01
  - Updated GPS::Point minimum required version to 0.10

0.09 2009-11-27
  - Added DateTime to handler output

0.08 2009-12-06
  - API changed between gpsd versions 2.90dev and 2.90
    - Added json=>true to ?WATCH request

0.07 2009-11-22
  - Support some changes in the protocol in 2.90dev.

0.06 2009-10-03T5:34:00Z
  - Satellite->reported property was dropped (now calculated)

0.05 2009-09-28T18:31:00Z
  - modified SKY->Satellites to reuse parent->constructor code.

0.04 2009-09-26T23:03:00Z
  - Documentation
  - Changed satelliteObjects to Satellites (still have satellites)
  - Changed DeviceObjects to Devices (still have devices)
  - Added Error trapping and better messages

0.03 2009-09-19T06:48:00Z
  - Fixed up DEVICE package
  - Fixed up ERROR package
  - Refactored a bunch of the code to make testing easier
  - Added more tests
  - Added Example

0.02 2009-09-16T00:20:00Z
  - Documentation
  - Updated SKY->satellites
  - Added SKY->satelliteObjects
  - Updated Makefile.PL -> prereq

0.01 2009-04-14T18:58:54Z
  - gpsd version 3.1 protocol


lib/Net/GPSD3/Return/SKY.pm
package Net::GPSD3::Return::SKY;
use strict;
use warnings;
use base qw{Net::GPSD3::Return::Unknown};
use DateTime;

our $VERSION='0.08';

=head1 NAME

Net::GPSD3::Return::SKY - Net::GPSD3 Return SKY Object

=head1 SYNOPSIS

=head1 DESCRIPTION

Provides a Perl object interface to the SKY object returned by the GPSD daemon.

An example JSON object:

  {
    "class":"SKY",
    "tag":"MID4",
    "device":"/dev/ttyUSB0",
    "time":1253593665.430,
    "hdop":23.60,
    "reported":9,
    "satellites":
      [
        {"PRN":15,"el":77,"az":123,"ss":0, "used":false},
        {"PRN":18,"el":25,"az":268,"ss":0, "used":false},
        {"PRN":27,"el":13,"az":150,"ss":0, "used":false},
        {"PRN":29,"el":47,"az":228,"ss":0, "used":false},
        {"PRN":5, "el":39,"az":58, "ss":46,"used":true },
        {"PRN":21,"el":41,"az":309,"ss":33,"used":true },
        {"PRN":10,"el":32,"az":61, "ss":40,"used":true },
        {"PRN":8, "el":12,"az":48, "ss":40,"used":true },
        {"PRN":2, "el":9, "az":124,"ss":0, "used":false}
      ]
  }

=head1 METHODS PROPERTIES

=head2 class

Returns the object class

=head2 string

Returns the JSON string

=head2 parent

Return the parent Net::GPSD object

=head2 device

=cut

sub device {shift->{"device"}};

=head2 tag

=cut

sub tag {shift->{"tag"}};

=head2 time

=cut

sub time {shift->{"time"}};

=head2 datetime

Returns a L<DateTime> object

=cut

sub datetime {
  my $self=shift;
  unless (defined($self->{"datetime"})) {
    $self->{"datetime"}=DateTime->from_epoch(epoch=>$self->time);
  }
  return $self->{"datetime"};
}

=head2 strftime

Returns the formatted datetime

=cut

sub strftime {
  my $self=shift;
  return $self->datetime->strftime($self->parent->strftime);
}

=head2 reported

Count of satellites in view

=cut

sub reported {
  my $self=shift;
  unless (defined($self->{"reported"})) {
    my @data=();
    push @data, $self->satellites;
    $self->{"reported"}=scalar(@data);
  }
  return $self->{"reported"};
}

=head2 used

Count of satellites used in calculation

=cut

sub used {
  my $self=shift;
  unless (defined($self->{"used"})) {
    my @data=();
    push @data, grep {$_->{"used"}} $self->satellites;
    $self->{"used"}=scalar(@data);
  }
  return $self->{"used"};
}

=head2 satellites

Returns a list of satellite data structures.

  my $satellites=$sky->satellites(); #[{},...]
  my @satellites=$sky->satellites(); #({},...)

=cut

sub satellites {
  my $self=shift;
  unless (ref($self->{"satellites"}) eq "ARRAY") {
    $self->{"satellites"}=[];
  }
  return wantarray ? @{$self->{"satellites"}} : $self->{"satellites"};
}

=head2 Satellites

Returns a list of L<Net::GPSD3::Return::Satellite> objects.

  my @satellites=$sky->Satellites; #(bless{},...)
  my $satellites=$sky->Satellites; #[bless{},...]
  
=cut

sub Satellites {
  my $self=shift;
  unless (defined($self->{"Satellites"})) {
    $self->{"Satellites"}=[
      map {$self->parent->constructor(%$_,
                                      class=>"Satellite",
                                      string=>$self->parent->encode($_))} 
        grep {ref($_) eq "HASH"} $self->satellites];
  }
  return wantarray ? @{$self->{"Satellites"}} : $self->{"Satellites"};
}

=head1 BUGS

Log on RT and Send to gpsd-dev email list

=head1 SUPPORT

Try gpsd-dev email list

=head1 AUTHOR

  Michael R. Davis
  CPAN ID: MRDVT
  STOP, LLC
  domain=>michaelrdavis,tld=>com,account=>perl
  http://www.stopllc.com/

=head1 COPYRIGHT

This program is free software licensed under the...

  The BSD License

The full text of the license can be found in the
LICENSE file included with this module.

=head1 SEE ALSO

L<Net::GPSD3>, L<DateTime>, L<Net::GPSD3::Return::Unknown>

=cut

1;
